src/models/search/Db.php

```diff
@@ -58,7 +59,8 @@ public function attributeLabels()
      */
     public function search($models)
     {
-        $dataProvider = new ArrayDataProvider([
+        $dataProvider = Yii::createObject([
+            '__class' => ArrayDataProvider::class,
```

Comment:
are we really discouraging the use of `new` in general? Does not sound right to me...
Also changes like this make the whole code depend on the `Yii` class directly. That should be avoided.

Reply:
I would love to know the way the development team recommends it. This is very important!
I will use it in subsequent PRs.

Reply:
And Event Model, etc.

Reply:
If there is no time to consider the merger, I will fix it later.

Reply:
@cebe No.
The point is that after removing `configure` feature from `BaseObject` passing `$config` array into constructor doesn't work. But it still works with `Yii::createObject`. But continue reading please.

So, previously it was:

```php
$object = new MyClass($param1, $param, [
    'prop1' => 'value1',
    'prop2' => 'value2',
    ...    
]);
```

With `Yii::createObject` it will look like this:

```php
$object = Yii::createObject(
    '__class' => MyClass::class,
    '__constructor()' => [
        'param1' => $param1,
        'param2' => $param2,
    ],
    'prop1' => 'value1',
    'prop2' => 'value2',
    ...    
]);
```

But I think the proper solution is to redo classes constructors to be usable in normal PHP way:

```php
$object = new MyClass($param1, $param, 'value1', 'value2');
$object->prop3 = 'value3';
]);
```

OR

```php
$object = (new MyClass($param1, $param, 'value1', 'value2'))->setProp3('value3');
```

I know that removing `configure` from BaseObject is a big change.
But it was most strange of Yii framework.
I think moving to normal constructors is a good way.

Reply:
>  But it was most strange of Yii framework.

You may find it strange but to me it was a very powerful feature, which I will be missing.

Consider the following example:

```php
$product = new Product([
    'scenario' => 'create',
    'name' => ".$tld Domain",
    'description' => "A domain name with ending $tld",
    'billable' => true,
    'auto_order' => false,
    'prepayed' => false,
    'price' => $this->faker->numberBetween(120, 5000),
    'interval' => 'YEAR',
    'object_type' => Domain::class,
    'object_match' => is_int($k) ? $tld : $k,
]);
```

It is hard to decide on whether any of these properties should be constructor arguments.
Even putting all required properties in the constructor makes the constructor argument list very long, passing values that way makes it unclear what these values are, e.g.:

```php
$product = new Product('create', ".$tld Domain", true, false, false, 'YEAR');
$product->setDescription("A domain name with ending $tld");
$product->setPrice($this->faker->numberBetween(120, 5000));
$product->setObjectType(Domain::class);
$product->setObjectMatch(is_int($k) ? $tld : $k);
```

Even putting all these into setters and having no constructor arguments has a lot of repetition in `->set...(...)`. To me such code looks very complex, when comparing it to an array configuration like above.

When a class has a few dependencies and only very few setters the array approach is not needed but if we have a lot of options, the array syntax makes the code much more readable.

Reply:
Other people's yii2.0 extension will be redesigned

Reply:
> You may find it strange but to me it was a very powerful feature, which I will be missing

Of course nobody should suffer from removing the feature.

1. For some classes configuration feature can be added, it is really easy. Base class or trait implementing this feature can be added.
2. `Yii::createObject` could be used to create objects from config and even more complex configs than available with original config array.

But I think classes should not have this feature by default. Most classes just don't need it.

Reply:
> But I think classes should not have this feature by default. Most classes just don't need it.

that sounds okay to me.

Reply:
@cebe internal object creation and application object creation are two different things. We've removed "configurable" for a reason of abusing it. It seriously violates encapsulation principle making code fragile. I see that in your example it becomes inconvenient but returning to array-configuration isn't the only option there. A good example are our migrations. First we've used arrays then switched to builder pattern:

```php
$product = (new Product(".$tld Domain"))
   ->scenario('create')
   ->description("A domain name with ending $tld")
   ->billable()
   ->noAutoOrder()
   ->notPrepaid()
   ->price($this->faker->numberBetween(120, 5000))
   ->interval('YEAR')
   ->type('Domain::class')
   ->match(is_int($k) ? $tld : $k);
```

Reply:
```php
new MyClass(false, false, 0, 1, 0, true, function() {}, 'bad')
```
Really bad

Reply:
The Yii 2.0 behavior is EXTREMELY useful and suits perfectly for data models, [DTO](https://en.wikipedia.org/wiki/Data_transfer_object), dependency-free and logic-free classes. 

As soon as logic or dependencies come, it [almost](https://github.com/yiisoft/yii2/blob/e16dcd967478ef682bbe21085ada589ffb158ad4/framework/grid/GridView.php#L276) [always](http://github.com/yiisoft/yii2/blob/e16dcd967478ef682bbe21085ada589ffb158ad4/framework/validators/CompareValidator.php#L104) [becomes](http://github.com/yiisoft/yii2/blob/e16dcd967478ef682bbe21085ada589ffb158ad4/framework/validators/DateValidator.php#L212) [a](http://github.com/yiisoft/yii2/blob/e16dcd967478ef682bbe21085ada589ffb158ad4/framework/validators/FileValidator.php#L165) [freaking](http://github.com/yiisoft/yii2/blob/e16dcd967478ef682bbe21085ada589ffb158ad4/framework/web/UrlManager.php#L171) [mess](http://github.com/yiisoft/yii2/blob/e16dcd967478ef682bbe21085ada589ffb158ad4/framework/filters/AccessControl.php#L98).

Once we want to utilize DI, the config passing [becomes](https://github.com/hiqdev/hipanel-module-finance/blob/8784bc9d7a9bbea157503ac824f7237368fd27e5/src/cart/storage/RemoteCartStorage.php#L56) a an abuse.

5 years ago DI was not very popular in PHP world, so creating everything with a `new` keyword was something perfectly good. Nowdays with developed DI, interoperability and general community skills growth we should not only follow the trends, but create them. The OLD parameters assigning way is great for some cases – we won't drop it, but we must clarify when it's a good idea and when it's not.

Reply:
We'll come to that, yes. Not right now though.

Reply:
@kids-return there's no one standard as @SilverFire, @cebe and me explained above.